### PR TITLE
Burndown Chart Feature

### PR DESCRIPTION
**Pull request by thutterer**
Requires phantomjs and redis.
Adding @kalabiyau for support.

### CODE REVIEW

File: app/models/scrum/burndown_generator.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      end
+    end
+
+    def render_and_upload
+      render_chart && upload_chart
+    end
+
+    def render_chart
+      phantom_output = []
+      Phantomjs.run("vendor/javascript/burndown-image.js", "#{url_prefix}/scrum/burndowns/#{@sprint_number}") { |line| phantom_output << line }
```
Comment (kalabiyau) [inline app/models/scrum/burndown_generator.rb]: Better to extract to method with Rails.root.join

File: app/models/scrum/burndown_generator.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    end
+
+    def upload_chart
+      burndown_card = Trello::Board.find(Rails.application.secrets.scc_current_board).lists.detect {|l| l.name == "Done Sprint #{@sprint_number}" }.cards.first
+      burndown_image = File.new("public/images/burndown-#{@sprint_number.to_s}.png")
+      burndown_card.add_attachment(burndown_image, name = 'burndown chart')
+    end
+
+  end
+end
```
Comment (kalabiyau) [inline app/models/scrum/burndown_generator.rb]: missing new line

File: app/views/layouts/application.html.haml [hunk 3]
```diff
@@ -3,6 +3,7 @@
     %title TeamDashboard
     %meta{:charset => "utf-8"}/
     %meta{:content => "width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no", :name => "viewport"}/
+    <script src='https://www.google.com/jsapi'></script>
```
Comment (kalabiyau) [inline app/views/layouts/application.html.haml]: I think you don't need to carry that in all the pages, better use it at the page it is used

File: app/controllers/api_controller.rb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+class ApiController < ActionController::Base
+
+end
```
Comment (kalabiyau) [inline app/controllers/api_controller.rb]: missing newline

File: app/controllers/scrum/burndowns_controller.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
+class Scrum::BurndownsController < ApplicationController
+
+  DAYS_PER_SPRINT = 10
+
+  def index
+    @sprints = Sprint.all
+  end
+
+  def show
+    @num = params[:number].to_i
```
Comment (kalabiyau) [inline app/controllers/scrum/burndowns_controller.rb]: That whole thing we need to wrap into a class, and to perform that heavy logic in the controller.

File: app/models/day.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class Day < ActiveRecord::Base
+  belongs_to :sprint
+
+  validates_presence_of :date
```
Comment (kalabiyau) [inline app/models/day.rb]: new and prefered way is validates :date, presence: true

http://guides.rubyonrails.org/active_record_validations.html

File: app/workers/burndown_worker.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class BurndownWorker
+  include Sidekiq::Worker
+  sidekiq_options retry: false
+
+  def perform(number)
+    Scrum::BurndownGenerator.new(number).render_and_upload
+  end
+end
```
Comment (kalabiyau) [inline app/workers/burndown_worker.rb]: missing newline

File: spec/factories/sprint.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :sprint, class: Sprint do
+    sequence(:number) { rand(99) }
+  end
+end
```
Comment (kalabiyau) [inline spec/factories/sprint.rb]: missing newline

File: spec/support/approvals.rb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+Approvals.configure do |c|
+  c.excluded_json_keys = {
+      :id =>/(\A|_)id$/,
+      :date => /_at$/
+  }
+end
```
Comment (kalabiyau) [inline spec/support/approvals.rb]: missing newline

**Comment by kalabiyau** [comment]
@thutterer please read about sidekiq and testing https://github.com/mperham/sidekiq/wiki/Testing
generally what we want to do for testing to not to use async_process, but rather inline process, so redis will not be needed

**Comment by thutterer** [comment]
> generally what we want to do for testing to not to use async_process, but rather inline process, so redis will not be needed


ok, I added the following to `spec/requests/burndown_generate_spec.rb`:
```ruby
require 'sidekiq/testing'
Sidekiq::Testing.inline!
```
running rspec now freezes when hitting that test. I guess because it is single-threaded but tries to render its own view. can we use Puma for this? Or what's the proper way for this situation?

**Comment by thutterer** [comment]
Oh, and also I just realized that running the tests does upload stuff to Trello. Good thing I still use a different board for testing ;) What's the solution here?

File: app/models/scrum/burndown_generator.rb (hunk 1 above)
Comment (thutterer) [inline app/models/scrum/burndown_generator.rb]: > Better to extract to method with Rails.root.join

Not sure what you mean by that.

**Comment by kalabiyau** [comment]
@thutterer 

>Oh, and also I just realized that running the tests does upload stuff to Trello. Good thing I still use a different board for testing ;) What's the solution here?

you can use VCR or webmock to stub external requests

**Comment by kalabiyau** [comment]
@thutterer so, now logical part of tests is green but check coverage

File: app/models/scrum/burndown_generator.rb (hunk 1 above)
Comment (kalabiyau) [inline app/models/scrum/burndown_generator.rb]:
```ruby
jsrunner_location = Rails.root.join "vendor/javascript/burndown-image.js"
sprint_url = "#{url_prefix}/scrum/burndowns/#{@sprint_number}"
Phantomjs.run(jsrunner_location, sprint_url) { |line| phantom_output << line }

```

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -3,7 +3,11 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
 
-  def index
+  before_action :get_sprint_data
```
Comment (kalabiyau) [inline app/controllers/application_controller.rb]: are you sure that every action in every controller derived from application_controller really need to load all the sprints every time?

File: app/controllers/scrum/api/burndowns_controller.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
+class Scrum::Api::BurndownsController < ApiController
+
+  def generate
+    sprint = Sprint.where(number: sprint_params[:meta][:sprint]).first_or_create
+    Day.from_params(sprint_params)
+    BurndownWorker.perform_async(sprint.number)
```
Comment (kalabiyau) [inline app/controllers/scrum/api/burndowns_controller.rb]: I would think of something like - moving that line to day after_update method, so every time day changed - we regenerate chart - which does makes sense almost all the time.

File: app/controllers/scrum/burndowns_controller.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
+class Scrum::BurndownsController < ApplicationController
+
+  def index
+    @sprints = Sprint.all
+  end
+
+  def show
+    num = params[:number].to_i
+    @sprint = Sprint.find_by_number num
```
Comment (kalabiyau) [inline app/controllers/scrum/burndowns_controller.rb]: find_by_number -> find_by(number: params[:number])

File: app/controllers/scrum/burndowns_controller.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+    redirect_to scrum_burndowns_url, flash: { error: "No data found for sprint #{num}"} and return unless @sprint
```
Comment (kalabiyau) [inline app/controllers/scrum/burndowns_controller.rb]: parentheses would not harm here to make it more readable - wrap scrum_burndowns_url

File: app/models/day.rb [hunk 14]
```diff
@@ -0,0 +1,24 @@
+class Day < ActiveRecord::Base
+  belongs_to :sprint
+
+  validates :date, presence: true
+  validates :date, uniqueness: true
+  validates :sprint, presence: true
+  validates :storypoints_total, presence: true
+  validates :storypoints_open, presence: true
+
+  def self.from_params(sprint_params)
```
Comment (kalabiyau) [inline app/models/day.rb]: :+1:

File: app/models/sprint.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
+class Sprint < ActiveRecord::Base
+  self.primary_key = :number
+  has_many :days, dependent: :destroy
+
+  validates_presence_of :number
+  validates_uniqueness_of :number
+
+  scope :latest, -> {order(:number).last}
```
Comment (kalabiyau) [inline app/models/sprint.rb]: again, latest and oldest :)

File: app/models/sprint.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class Sprint < ActiveRecord::Base
+  self.primary_key = :number
+  has_many :days, dependent: :destroy
+
+  validates_presence_of :number
```
Comment (kalabiyau) [inline app/models/sprint.rb]: validates_presence_of -> validates :number, presence: true

File: app/views/layouts/application.html.haml [hunk 17]
```diff
@@ -40,11 +40,12 @@
               %i.fa.fa-bar-chart>
               Timeline
             .list-selected
-          %li.submenu-entry.scrum{ data: {submenu: 'scrum'} }
-            = link_to '/scrum/burndown' do
-              %i.fa.fa-fire>
-              Burndown
-            .list-selected
+          - if @sprints.any?
```
Comment (kalabiyau) [inline app/views/layouts/application.html.haml]: :+1:

File: app/views/scrum/burndowns/index.html.slim [hunk 18]
```diff
@@ -0,0 +1,11 @@
+h2 Scrum Burndown Charts
+
+.row
+  - @sprints.each do |sprint|
+    .col-sm-6
+      a href=scrum_burndown_url(sprint.number)
+        .panel.panel-default
+          .panel-heading
+            = "Sprint #{sprint.number}"
```
Comment (kalabiyau) [inline app/views/scrum/burndowns/index.html.slim]: Perfect candidate to be moved to decorator

File: app/views/scrum/burndowns/index.html.slim [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+          .panel-body
+            img src="/images/burndown-#{sprint.number}.png" class='img-responsive'
```
Comment (kalabiyau) [inline app/views/scrum/burndowns/index.html.slim]: either should be in a helper or doen image_tag

File: app/views/scrum/burndowns/show.html.slim [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+  .panel-body
+    #spinner.text-center
+      i.fa.fa-5x.fa-cog.fa-spin
+    = javascript_include_tag '//www.google.com/jsapi'
+    #chart
+    = render_chart(@chart, 'chart')
+
+.text-center
+  /prev button
+  - if @sprint.number > Sprint.oldest.number
```
Comment (kalabiyau) [inline app/views/scrum/burndowns/show.html.slim]: That does make sense to look into kaminari or will_paginate

File: app/views/scrum/burndowns/show.html.slim [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+  /index button
+  a.btn.btn-default role="button" href=scrum_burndowns_url
+    i.fa.fa-list-ol
+  /next button
+  - if @sprint.number < Sprint.latest.number
+    a.btn.btn-default role="button" href=scrum_burndown_url(@sprint.number + 1)
+      i.fa.fa-chevron-right
+  - else
+    a.btn.btn-default.disabled role="button" href="#"
+      i.fa.fa-chevron-right
```
Comment (kalabiyau) [inline app/views/scrum/burndowns/show.html.slim]: missing new lie

File: lib/trello_wrapper.rb [hunk 22]
```diff
@@ -35,6 +35,18 @@ def selectable_cards
       return {}
     end
 
+    def done_list sprint_number
+      Trello::Board.find(Rails.application.secrets.scc_current_board).lists.detect {|l| l.name == "Done Sprint #{sprint_number}" }
+    rescue Trello::Error
```
Comment (kalabiyau) [inline lib/trello_wrapper.rb]: Trello::Error is too broad I think. What if access is lost or some other error will happen - you will never know about it

**Pull request merged**